pkg/metrics/metrics.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/apimachinery/pkg/types"
+	k8smetrics "k8s.io/component-base/metrics"
+	"k8s.io/klog"
+)
+
+const (
+	labelDriverName               = "driver_name"
+	labelOperationName            = "operation_name"
+	labelOperationStatus          = "operation_status"
+	snapshotController            = "snapshot_controller"
+	operationLatencyMetricName    = "operation_total_seconds"
+	operationLatencyMetricHelpMsg = "Total number of seconds spent by the snapshot controller for an operation to complete end to end"
+)
+
+type OperationState string
+
+const (
+	// InterimFailure is an operation state which means the controller will keep
+	// on retrying the operation upon encountering this type of error.
+	InterimFailure OperationState = "InterimFailure"
+
+	// TerminatingFailure is an operation state which means the controller has encountered
+	// an error which it's not able to resume the operation and has marked the
+	// permanently failure of the operation.
+	TerminatingFailure OperationState = "TerminatingFailure"
```

Comment:
How do you differentiate a TerminatingFailure vs a InterimFailure?

Reply:
in controller, some errors, for example controller failure or timeout when creating snapshot, are considered to be interim and controller will reconcile later to retry. Basically, if a controller encountered an error and will try to reconcile later should be considered as interim failure, otherwise should be terminating error. WDYT?

Reply:
Sure.  We added a isFinalError() in the CreateSnapshot timeout PR, but that depends on gRPC error codes.  We can probably do something similar for the snapshot controller.

https://github.com/kubernetes-csi/external-snapshotter/pull/261/files#diff-de5dbd65778c167a3d05cd17929d6851R658-R681

Reply:
What is an example of a terminating failure?

Reply:
for example when an operation timed out from CSI, controller failed to update status field of an object due to API error etc.

Reply:
For example, CSI returned a failure which is not recoverable from the controller automatically.

Reply:
Are there any error codes returned by the CSI driver that the controller will not retry?How does "user canceled" get counted?

Reply:
Snapshot controller does not directly get CSI driver error codes.  For create snapshot, maybe we can check the VolumeSnapshotBeingCreated annotation on the content?  If create snapshot failed and content does not have the annotation, that means it is a terminating failure.

Reply:
But I thought we retry on any error until the object is deleted?

Reply:
Yes.  I just wonder if we need to change the logic now that we are talking about terminating failure.

Reply:
My take on that would be sidecar controller *could* update the content object with a permanent error in the status on those cases instead of utilizing annotation.

Reply:
I'm confused. I thought we always retry no matter the error. So the only time we stop retrying is when the user deletes the VolumeSnapshot object.

Reply:
Yes.  Controller will always retry to create the snapshot no matter what kind of error it is.  Now I think we don't need to track the interim failure such as timeout.  We only need to track a terminating failure, i.e., driver finally finished the create snapshot call with a final error which is not a timeout.  Controller will still retry later until the user deletes the VolumeSnapshot object.  If the controller can never successfully create the snapshot, then the failure count will continue to increment with each retry.